File: Hara-Chap04.r
library(tidyverse)
library(ggExtra)

# カレントディレクトリの設定
# setwd("/Users/soichi/Dropbox/Programing")

# csvデータの読み込み
input_data <- read.csv("carprice.csv")

# データの先頭10行を確認する
head(input_data,10)
# データの構造を確認する
str(input_data)
# データの概要を確認する
summary(input_data)

# データの各種統計量をみる。
mean(input_data$price)
median(input_data$price)
var(input_data$price)
sd(input_data$price)
max(input_data$price)
min(input_data$price)

# dplyrをつかってグループ別統計量を見る。
summary_price <- input_data %>% 
  dplyr::group_by(doornumber) %>%
  dplyr::summarise(
    mean_price = mean(price),
    median_price = median(price),
    var_price = var(price),
    sd_price = sd(price),
    max_price = max(price),
    min_price = min(price)
  )
print(summary_price)


# ggplotでヒストグラム
g <- ggplot(input_data, aes(x=price)) + geom_histogram(boundary = 0, binwidth=5000) 
g <- g +  scale_x_continuous(breaks=seq(5000, 50000, by=5000)) 
g <- g + theme_bw() + xlab("price") + ylab("count") 
print(g)

# 箱ひげ図
## 基本関数
boxplot(input_data$price)
## ggplot
ggplot(input_data, aes(y = price)) + geom_boxplot()

# 散布図
## 基本関数
plot(input_data$enginesize, input_data$price)
## ggplot
g <- ggplot(input_data,aes(x=enginesize,y=price)) + geom_point(alpha = 0.3)
g <- g + ggtitle("Engine and Price") + xlab("Engine") + ylab("Price")
print(g)

# 共分散
cov(input_data$price,input_data$enginesize)
# 相関係数
cor(input_data$price,input_data$enginesize)
# 相関係数行列
dat <- input_data %>% 
  dplyr::select(price,enginesize) #priceとenginesizeを取り出す
cov(dat)
cor(dat)

# 散布図とヒストグラム
g <- ggplot(input_data, aes(x = enginesize, y = price)) + geom_point()
ggMarginal(g, type = "histogram", bins = 20,margins = "both", size = 5)

# 複数変数間の情報量が多い図
dat2 <- input_data %>% 
  dplyr::select(price, enginesize, wheelbase, horsepower, carheight)
# install.packages("psych")
library(psych)
psych::pairs.panels(dat2,hist.col="white", lm=T, rug = F, ellipses=F)

# 相関係数を色で表す
# install.packages("ggcorrplot")
library(ggcorrplot)
res <- cor(dat2)
g <- ggcorrplot(corr = res, hc.order = TRUE, method = "square", lab = TRUE)
print(g)
# macだとggsaveは使えないので，工夫が必要。
ggsave(file = "cor_heat.png", plot = g)

# おしまい

File: Hara-Chap04.py
import pandas as pd
import io
import os
import matplotlib.pyplot as plt

# カレントディレクトリの場所を確認
os.getcwd()
# csvデータの読み込み
input_data = pd.read_csv("carprice.csv")

# データの先頭10行を表示 head()
input_data.head(10)

# データの概要を表示 info()
input_data.info()

# データの記述統計をまとめて表示 describe()
input_data.describe()

# 各種記述統計量を計算 mean()
## 平均
print(input_data.price.mean())
## 以下同様

# pandasのgroupbyでカテゴリ別に統計量を計算 groupby
## 平均
print(input_data.groupby("doornumber")["price"].mean())
## 以下同様

# ヒストグラム hist()
## 価格のヒストグラム
plt.hist(input_data.price)
plt.xlabel('price') # x軸のラベル
plt.ylabel('count') # y軸のラベル
plt.grid(True) # 補助線を引く

# 箱ひげ図 boxplot
plt.boxplot(input_data.price)
plt.grid(True)

# 4.5 データを把握する 変数間の関係性

# scipyパッケージを読む
import scipy.stats
# エンジンサイズと価格の散布図
plt.scatter(input_data.enginesize, input_data.price, s = 50, c = "blue", alpha = "0.3", linewidths = 1, edgecolors = "blue")
plt.title('Engine and Price')
plt.ylabel('Price')
plt.xlabel('Engine')

# 共分散や相関行列

# 科学技術計算ライブラリ numpyを読み込む
import numpy as np
#共分散 cov
print(np.cov(input_data.price, input_data.enginesize))
#相関行列 corrcoef
print(np.corrcoef(input_data.price, input_data.enginesize))

# 散布図とヒストグラムを書く
import seaborn as sns
sns.jointplot(input_data.price, input_data.enginesize)

# 5つのをまとめてリストx_listに格納しておく
x_list = pd.concat([input_data.price, input_data.enginesize, input_data.wheelbase, input_data.horsepower, input_data.carheight], axis = 1)
# pairplot 
sns.pairplot(x_list, kind = "reg")

# 相関係数をプロットし，色で強弱を示す。
plt.figure(figsize = (12, 9)) # 図の大きさ
sns.heatmap(x_list.pct_change().corr(), annot = True, cmap = "Blues")

# 作業結果の保存
sns.pairplot(x_list, kind = "reg").savefig("test.png")
